LibraBFT/Impl/Consensus/EpochManager/Properties.agda

```diff
@@ -0,0 +1,257 @@
+
+  Some possibilities to consider:
+
+  1.  It doesn't matter that the invariant (as currently stated) does not hold at this point.
+      This could be because:
+
+      1a. it gets restored somewhere downstream before any damage occurs, or
+
+      1b. no damage will occur (e.g, perhaps when considering whether to vote, we will check
+      sdLastVote, find that it is for the wrong epoch, and bail out or repair the invariant first)
+
+  2.  It does matter.  Some possibilities:
+
+      2a. We could change the code to ensure the invariant (as currently stated) is not falsified.
+          I think this could be acheived by setting sdLastVote to nothing in the case in which line
+          289 CHANGES sdEpoch (it is probably important NOT to set sdLastVote to nothing if line 289
+          happens to overwrite the SAME value to sdEpoch)
+
+      2b. We could confirm that we are in case 1, or possibly change the code to make sure we are.
+          In this case, we will need to weaken the invariant so that we can prove it.  It might be
+          that changing SafetyDataInv to be something like:
+
+             Meta.getLastVoteEpoch sd ≡ sd ^∙ sdEpoch
+           → Meta.getLastVoteRound sd ≤ sd ^∙ sdLastVotedRound
+
+          in case we confirm/ensure that we are in case 1b.
+
+      2c. It could be that the implementation maintains an invariant that sdLastVote of the
+          SafetyData within the EpochManager's SafetyRulesManager is always nothing.  This is the
+          case for the INITIAL EpochManager, constructed using EpochManager.new (see line 42 in
+          SafetyRules, which is reached 
+
+  In any case, we need to understand more detail to determine what to do.  I have written the above
+  to facilitate discussion.
+
+  One thing to note is that for the first epoch, sdLastVote is nothing (startConsensus ->
+  startForTesting -> MockSharedStorage.new*; see MockSharedStorage, line 19).  Therefore, we could
+  prove a weaker version of startRoundManagerSpec.contract', which assumes that sdLastVote is
+  nothing, and this would suffice for continuing our current work on proving safety for the first
+  epoch.
+
+  TENTATIVE CONCLUSION
+
+  Writing the last paragraph above caused me to challenge my assumption that sdLastVote in the
+  SafetyRules obtained from the EpochManager's SafetyRulesManager might not be `nothing`.  In
+  particular, it seems that when we start a new epoch we ensure that it IS `nothing` (startNewEpoch
+  -> expectNewEpoch -> startProcessor -> startForTesting -> MockSharedStorage.new*; see
+  MockSharedStorage, line 19); note that the EpochManager returned by startProcessor is the one
+  provided to the PMNewEpochManager constructed by expectNewEpoch).  Therefore, I have strengthened
+  the initial EpochManager invariant accordingly, also strengthened performInitializeSpec, and used
+  these to complete the proofs in startRoundManagerSpec.  After review I will remove this comment,
+  assuming I got this right.
+
```

Comment:
Would you like me to step through the Haskell code (in my head) to determine "for sure" what it does with `sdLastVote`?

Reply:
Thank you, it would be great to get confirmation that `sdLastVote` is `nothing` in the `EpochManager` created at epoch change.  It certainly makes sense, and is true as far as I can see.

Reply:
OK - Will do.

Reply:
DONE: sdLastVote IS set to Nothing on EpochChange.

FWIW:  A new `RoundManager` is created at epoch change (not a new `EpochManager`).

Reply:
Great, thanks.  I will update the comments and merge.